### DISCUSSION TITLE
ControlTheme Changes

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#10391] [General]
I'm looking at the new ControlTheme changes. It is not yet documented. My initial experience is that it breaks my existing user controls; `PART_Track` in a Slider is not found and thus the control doesn't work. Probably it now has to be inside a ControlTheme now to be found?

I wanted to talk about the philosophy of the new structure.

What are the benefits?

I'm looking at [FluentSlider](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Slider.xaml), and it defines FluentSliderHorizontalRepeatButton, FluentSliderVerticalRepeatButton, FluentSliderThumbTheme, and {x:Type Slider} (that's default theme I suppose?)

From what I observe, ControlTheme is a replacement for Style Classes, except that it doesn't have a hierarchical structure and the naming is flat? Is the naming prone to collisions between 3rd party controls? What happens if 2 separate libraries have "MediaSeekbar" theme?

From what I understand, the difference is that we now set the styles with Theme instead of Classes property?

Is the API considered stable at this point, or prone to further changes?

How do we adapt to Light vs Dark theme?

In my case of [MediaPlayerUI](https://github.com/mysteryx93/MediaPlayerUI.NET/tree/master/Avalonia/Styles), I would like to make various skins available, in Light and Dark flavors. What would be the best way of structuring my code compared to what I have now? Currently, you include the Theme and the Colors files that you want; but switching them can be a bit more complex. In practice; my existing app supported Dark Theme but the MediaPlayer doesn't respect that.

Finally, FluentAvalonia used to have a property to exclude unused control styles to reduce memory usage and improve loading time. That was removed. Does the new Avalonia remove the load benefit of having less controls, or that's still like before?

Another thing. I have `<Style Selector="Button.toolbar, ToggleButton.toolbar">`... ControlTheme cannot have multiple `TargetType`, do I have to have 2 copies of the same code? And both copies would need separate keys? Like `ToolbarButton` and `ToolbarToggleButton`? That gets a bit ugly if that's the case.

And one more difference. I had to change the include of my styles to this; but it now complains that it cannot resolve the path. Do we still need to manually include styles and themes for 3rd party controls, and how do we do that now?

```xaml
<Application.Styles>
    <FluentTheme />
</Application.Styles>
<Application.Resources>
    <ResourceDictionary>
        <ResourceDictionary.MergedDictionaries>
            <MergeResourceInclude Source="avares://MediaPlayer.Avalonia/Styles/Default/Theme.axaml" />
            <MergeResourceInclude Source="avares://MediaPlayer.Avalonia/Styles/Colors/Gray.axaml" />
        </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
</Application.Resources>
```

**Comment by timunie** [comment]
It's not a replacement of Styles for now. styles are styll there and will stay. The benefit of ControlTheme is you can entilly replace them or use BasedOn. This wasn't possible before. Also you can apply a complete different ControlTheme on a specific Window for example. 

Also there might be generics in the future, so the StyleInclude is not needed anymore. 

I think all controls with a PART prefix must be present in your ControlTemplate. Whether it is inside a Style or ControlTheme is not important. 

Changing to Light or Dark is done via RequestedThemeMode iirc. This can be done on any control or for the complete App.

I hope some of your questions become clearer now.

**Comment by amwx** [comment]
> I wanted to talk about the philosophy of the new structure.
> 
> What are the benefits?

For that I'd say go read the previous discussions here on github about ControlThemes. They're quite extensive

> I'm looking at the new ControlTheme changes. It is not yet documented. My initial experience is that it breaks my existing user controls; `PART_Track` in a Slider is not found and thus the control doesn't work. Probably it now has to be inside a ControlTheme now to be found?

Clarification: Is this after you tried to rewrite your styles using control themes, or just your existing styles are conflicting with the ControlThemes? `ControlTheme` works closer to WPF/UWP `Style`s, but the existing `Selector` system still works and has priority over `ControlTheme`. 


> I'm looking at [FluentSlider](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Slider.xaml), and it defines FluentSliderHorizontalRepeatButton, FluentSliderVerticalRepeatButton, FluentSliderThumbTheme, and {x:Type Slider} (that's default theme I suppose?)

For external customization, different template parts can have different Themes applied, just like WPF. In `Slider`'s case, the first two are the themes for the repeat buttons in the template, the third is the thumb, and the `{x:Type Slider}` says that that should be the default theme loaded `Slider`. If you want alternate themes, you can change that to a string key like the other three you listed and then reference that:
```Xaml
<ControlTheme TargetType="Slider" x:Key="MySliderTheme" />

<Slider Theme="{StaticResource MySliderTheme}" />
```

`Theme` could be set in `Selector` too, so you could apply existing style classes or pseudoclasses that change and set the `Theme` property too.

 
> How do we adapt to Light vs Dark theme?

Typically light vs dark theme is just changing colors/brushes, the template itself doesn't usually change, in which case that's where you'll want to look at the new `ThemeDictionary`. If the template changes too, that's more difficult, but an existing Selector might be able to take care of it as a child of the ControlTheme `Selector="^[ActualThemeVariant=Dark]"` (where `^` means take the parent Style or ControlTheme


> Finally, FluentAvalonia used to have a property to exclude unused control styles to reduce memory usage and improve loading time. That was removed. Does the new Avalonia remove the load benefit of having less controls, or that's still like before?

Resources are now lazily loaded the first time a reference to them is taken, thus there was no need to carry this option forward. Since `ControlTheme` is declared in a `ResourceDictionary`, it's a resource and gets lazily evaluated, so this benefit now applies here too.

> 
> Another thing. I have `<Style Selector="Button.toolbar, ToggleButton.toolbar">`... ControlTheme cannot have multiple `TargetType`, do I have to have 2 copies of the same code? And both copies would need separate keys? Like `ToolbarButton` and `ToolbarToggleButton`? That gets a bit ugly if that's the case.

I think `ToggleButton` inherits from `Button`, so you can declare a common control theme:
```Xaml
<ControlTheme TargetType="Button" x:Key="ToolbarButtonTheme" />
```

and use that (I think this will work, I haven't tested it)
```Xaml
<Button Theme="ToolbarButtonTheme" />
<ToggleButton Theme="ToolbarButtonTheme" />
```

Though keep in mind togglebutton has extra visual states (:checked, etc) that a normal button does not.


> And one more difference. I had to change the include of my styles to this; but it now complains that it cannot resolve the path. Do we still need to manually include styles and themes for 3rd party controls, and how do we do that now?
> 
> ```
> <Application.Styles>
>     <FluentTheme />
> </Application.Styles>
> <Application.Resources>
>     <ResourceDictionary>
>         <ResourceDictionary.MergedDictionaries>
>             <MergeResourceInclude Source="avares://MediaPlayer.Avalonia/Styles/Default/Theme.axaml" />
>             <MergeResourceInclude Source="avares://MediaPlayer.Avalonia/Styles/Colors/Gray.axaml" />
>         </ResourceDictionary.MergedDictionaries>
>     </ResourceDictionary>
> </Application.Resources>
> ```

That seems to be a bug with `MergeResourceInclude`. It still works. If the error bothers you, just use a normal `ResourceInclude`

**Comment by mysteryx93** [comment]
> That seems to be a bug with MergeResourceInclude. It still works. If the error bothers you, just use a normal ResourceInclude

Leaving MergeResourceInclude doesn't allow me to launch; replacing it with ResourceInclude works. And Gray.axaml must remain in Styles. Now the app launches.

Found the problem with Minimum, Maximum and Value not being set on the Track.

Found this at the bottom of the theme file. That's new. Copied it and now it works.
```xaml
   <Style Selector="^ /template/ Track#PART_Track">
      <Setter Property="Minimum" Value="{TemplateBinding Minimum}" />
      <Setter Property="Maximum" Value="{TemplateBinding Maximum}" />
      <Setter Property="Value" Value="{TemplateBinding Value, Mode=TwoWay}" />
    </Style>
    <Style Selector="^ /template/ Border#TrackBackground">
      <Setter Property="BorderThickness" Value="2" />
      <Setter Property="BorderBrush" Value="{DynamicResource ThemeBorderLowBrush}" />
    </Style>
    <Style Selector="^ /template/ TickBar">
      <Setter Property="Ticks" Value="{TemplateBinding Ticks}" />
    </Style>
    <Style Selector="^:disabled /template/ Grid#grid">
      <Setter Property="Opacity" Value="{DynamicResource ThemeDisabledOpacity}" />
    </Style>
```

Almost. Now it launches but crashes with

> Cannot set direct property 'Minimum' in 'Slider:vertical /template/ Track#PART_Track' because the style has an activator.

What does that mean?

**Reply by rabbitism** [reply]
DirectProperty cannot be set in styles.

**Reply by mysteryx93** [reply]
well that's a copy/paste from the default style

**Reply by rabbitism** [reply]
I'm checking with my phone. Will take a look later.

**Reply by rabbitism** [reply]
@mysteryx93 I think the default theme have max/min bindings in control template, but not style setter.

**Reply by mysteryx93** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Simple/Controls/Slider.xaml#L112

**Reply by mysteryx93** [reply]
Found my issue: Style was within another Style. Moved it out and it works.

**Comment by mysteryx93** [comment]
Another problem with this
```xaml
    <Style Selector="^[IsPlaying=True] Button#PlayButton /template/ ContentPresenter">
        <Setter Property="Content">
            <Template>
                <Image Source="/Styles/Default/Icons/pause.png" />
            </Template>
        </Setter>
    </Style>
</ControlTheme>
```

> ControlTheme style may not directly contain a child or descendent selector.

**Reply by maxkatz6** [reply]
For the users of this control, it would be unexpected if control changes style of the content user put there.
In your case, style assumes there is a user content Button#PlayButton as a logical child of your control.
Also, limited perf-wise.


Checked this source code, it seems like PlayButton is part of the template
https://github.com/mysteryx93/MediaPlayerUI.NET/blob/c5907603e8373ee824e88b5adfbc0e044b94be12/Avalonia/Styles/Default/Theme.axaml#L331

So, should it be `Selector="^[IsPlaying=True] /template/ Button#PlayButton /template/ ContentPresenter"` ?
Which also won't be possible to change inner templates with control themes. But you can do something like this:
```xaml
<Style Selector="^[IsPlaying=True] /template/ Button#PlayButton">
    <Setter Property="Theme" Value="{StaticResource MediaToolbarPlayingButton}" />
</Style>
```

**Reply by mysteryx93** [reply]
Having to create a separate Theme just for playing state? A toggle button with 2 contents based on a boolean is a pretty common scenario. There's no cleaner way?

**Reply by rabbitism** [reply]
> Having to create a separate Theme just for playing state? A toggle button with 2 contents based on a boolean is a pretty common scenario. There's no cleaner way?

I would suggest use ToggleSwitch for this scenario, it holds Content, OnContent and OffContent, you can switch the visibility of them on different state.

**Reply by maxkatz6** [reply]
@mysteryx93 just in case, if doesn't have to be a whole new theme. You can use BasedOn and add a small style to it. 
Or use ToggleButton/ToggleSwitch from the beginning. As this control has multiple states build in. 
Simple button doesn't (aside from pointerover and pressed). 

Also, old styles do work too, but it needs to be defined outside of the control theme scope.

**Reply by timunie** [reply]
@mysteryx93 another idea would be to add both contents and bind to IsVisible. You can use the `!` operator to negate the boolean

**Reply by mysteryx93** [reply]
The reason I wasn't using a ToggleButton for Play/Pause button was because the decision as to whether it is playing comes from the business layer, not from the UI, thus I can't directly bind the IsPlaying property two-way.

Setting both contents with IsVisible actually solves the problem in a very simple way! Thanks

There are quire a few differences between standard styles and ControlTheme. There still remains visual differences for me to figure out. It could be good to have a document listing all differences that one can expect and how to solve them.

**Comment by mysteryx93** [comment]
Seeing another change.

Here the Repeater has a custom template defined.
https://github.com/mysteryx93/MediaPlayerUI.NET/blob/master/Avalonia/Styles/Default/Theme.axaml#L48

It now works perfectly when using FluentAvalonia; but when using FluentTheme, there's some inherited stuff that still gives it some funky behavior when clicking, even though I'm defining the whole template.

**Comment by mysteryx93** [comment]
[MediaPlayer.Avalonia](https://github.com/mysteryx93/MediaPlayerUI.NET) now works when using SimpleTheme or FluentAvalonia; but when using FluentTheme, something gets inherited that causes visual glitches when clicking on the track. What could that be? I'm unable to catch a DevTools snapshot while clicking so it's hard to know what causes that visual glitch or where it's defined.

And what about HighContrast theme, is that still natively supported?
